specs/interop/superchain_backend.md

```diff
@@ -0,0 +1,156 @@
+
+## Message Safety
+
+This method MUST enforce the [Identifier](./messaging.md#message-identifier)'s chainId is present in the chain's
+[dependency set](./dependency_set.md).
+
+This method MUST enforce the attributes listed in the [Identifier](./messaging.md#message-identifier) matches the log
+and block attributes returned by peer's `eth_getLogs` request with the `blockNumber` and `logIndex` parameters.
+
+This method MUST enforce the provided payload matches the bytes computed when
+constructing the [MessagePayload](./messaging.md#message-payload) against the fetched log.
+
+This method MUST return the applicable [safety label](./verifier.md#safety) for the message, or `Invalid` upon failures.
+
+This method MUST NOT enforce the [timestamp invariant](./messaging.md#timestamp-invariant) on the provided messages. The
+block builders and verifiers must locally do so based on their configured level of safety.
+
+## Message Safety Resolution
+
+### Unsafe to Cross-Unsafe
+
+In order to discern between unsafe and cross-unsafe blocks, the backend maintains a graph of
+dependencies between blocks can include interop messages between them.
+
+We describe a psuedo-implementation of cross-unsafe message resolution using this graph. Every node in
+this graph represents a block with edges representing the dependent relationship between them. Upon
+receipt of unsafe payloads, the graph is extended with edges added on detection of interop transactions.
+
+```python
+## Graph Nodes
+blocks = dict() # (hash) -> block
+number_to_block_hash = dict() # (chain_id, int) -> hash
+latest_blocks = dict() # (chain_id) -> hash
+
+## Graph Edge
+##   - The initiating side is represented with the block number since the corresponding payload may
+##   not have been seen at the time the executing message has been processed.
+dependents = dict() # (initiating_chain_id, block_number) -> List[(executing_chain_id, block_hash)]
+dependencies = dict() # (executing_chain_id, block_hash) -> List[(initiating_chain_id, block_number)]
+
+class Block:
+    chain_id: int
+    number: int
+    parent: Block
+
+    executing_messages: List[(MsgId, MsgPayload)]
+    unverified_msg_ref_count: int
+
+def block_safety(chain_id, block_hash):
+    block = blocks[block_hash]
+    if block.unverified_msg_ref_count > 0:
+        return UNSAFE
+
+    ## All messages were validated but we also need to ensure all
+    ## transitive block dependencies have had their messages also
+    ## validated
+    for initiating_chain_id, block_number in dependencies[(chain_id, block_hash)]:
+        if block_bumber > latest_heights[tx.msg_origin]:
+            return UNSAFE
+
+        dependency_block = blocks[number_to_block_hash[(initiating_chain_id, block_number)]]
+        if block_safety(initiating_chain_id, dependency_block.hash) == UNSAFE:
+            return UNSAFE
+
+    return CROSS_UNSAFE
+
+def add_unsafe_payload(chain_id, payload):
```

Comment:
Can you add a comment on each of these functions that states how its called? ie what triggers it being called